graphix/linalg_validations.py

```diff
@@ -2,29 +2,58 @@
 
 from __future__ import annotations
 
+import functools
+from typing import TYPE_CHECKING, TypeVar
+
 import numpy as np
+import numpy.typing as npt
+from typing_extensions import Concatenate, ParamSpec
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+
+_T = TypeVar("_T", bound=np.generic)
+_P = ParamSpec("_P")
+
+
+def is_square(matrix: npt.NDArray[_T]) -> bool:
+    """Check if matrix is square."""
+    if matrix.ndim != 2:
+        return False
+    rows, cols = matrix.shape
+    return rows == cols
+
+
+def _is_square_deco(
+    f: Callable[Concatenate[npt.NDArray[_T], _P], bool],
+) -> Callable[Concatenate[npt.NDArray[_T], _P], bool]:
+    """Check if matrix is square, and then call the function."""
+
+    # wraps(_f) has very messy annotation, but still callable
+    @functools.wraps(f)
+    def _f(matrix: npt.NDArray[_T], *args: _P.args, **kwargs: _P.kwargs) -> bool:
+        if not is_square(matrix):
+            raise ValueError("Need to ensure that matrix is square first.")
+        return f(matrix, *args, **kwargs)
+
+    return _f  # type: ignore[return-value]
 
 
-def check_square(matrix: np.ndarray) -> bool:
+@_is_square_deco
+def is_qubitop(matrix: npt.NDArray[_T]) -> bool:
```

Comment:
I would prefer that `is_qubitop` returns `False` if `matrix` is not square instead of raising a `ValueError`. We could define `assert_qubitop` that raises `ValueError` with the respective messages if the matrix is not square or if the size is not a power of two.

The same comment applies to other `is_*` predicates.

Reply:
I don't want both `is_` and `assert_` to be in the pkg. as they're essentially the same and `is_` functions are mainly used internally.

> `is_qubitop` returns `False` if `matrix` is not square

For this suggestion, I'm for that.

Reply:
Resolved in bb8d30f8babbd3b6adb3137717ddeedd4a8771d9 .